### PR TITLE
Bowling project - both tests and code

### PR DESCRIPTION
**Pull request by Ruthietta**
Authors: @NG90 @DamiXioR @WBigaj @Marimeiko

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
Wow. Very good job ;~}

File: bowling/docs/ProjectPlan.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Bowling - project plan
```
Comment (ziobron) [inline bowling/docs/ProjectPlan.md:2]: You have documentation! That's great 💚

File: bowling/docs/ProjectPlan.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+- **isHelpToBePrinted()** public function - it checks if “--help” or “-h” is in input arguments.
+- **printHelp()** public function - prints program help
+- **calculateScores()** public function - create object of class DirectoryHandler with directory name read from ArgumentParser object. For every file returned by DirectoryHandler object, create Lane class object and store it in vector.
+- **getOutputFileName()** private function - returns output file name (if exists) from ArgumentParser
+- **outputScores()** public function - use checkOutputFile() and showScores() functions always and saveScores() function depends on checkOutputFile() return
+- **isOutputFileGiven()** private function - checks if the output file was given
+- **showScores()** private function - show on screen scores
+- **saveScores()** private function - if function checkOutputFile() returns true, create an object of class FileHandler for output and write scores to the file by calling operator<< on every Lane object.
```
Comment (ziobron) [inline bowling/docs/ProjectPlan.md]: we generally don't put private functions into the documentation. They are subject to change more often than the public interface.

File: bowling/docs/plant_uml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+class FileHandler{
+-std::fstream fileStream_
+-std::string fileName_
+
++FileHandler(std::string fileName, enum FileAccess fileAccess)
++~FileHandler()
++std::string readLine()
++bool write(std::string)
+}
+@enduml
```
Comment (ziobron) [inline bowling/docs/plant_uml:102]: 💚

File: bowling/src/argumentParser.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    return parsedArguments_[static_cast<size_t>(argumentNumber)];
+}
+
+bool ArgumentParser::checkIfArgumentExists(const std::string& argument) const
+{
+    auto foundIt = std::find(parsedArguments_.begin(), parsedArguments_.end(), argument);
+    if (foundIt != parsedArguments_.end()) {
+        return true;
+    }
+    return false;
```
Comment (ziobron) [inline bowling/src/argumentParser.cpp]:
```suggestion
    return foundIt != parsedArguments_.end();
```
simplification

File: bowling/src/argumentParser.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            }
+        }
+        parsedArguments_.shrink_to_fit();
+        argumentsNumber_ = (size_t)(argc - 1);
+    }
+}
+
+std::string ArgumentParser::getArgument(const ArgumentNumber argumentNumber) const
+{
+    if (static_cast<size_t>(argumentNumber) >= argumentsNumber_) {
```
Comment (ziobron) [inline bowling/src/argumentParser.cpp]: you can extract a varaible, it's used twice
```suggestion
    size_t argNumber = static_cast<size_t>(argumentNumber);
    if (argNumber >= argumentsNumber_) {
```

File: bowling/tests/argumentParser.ut.cpp [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+using ArgNumber = ArgumentParser::ArgumentNumber;
+
+SCENARIO("program arguments and their number should be returned after parsing", "argumentParser")
+{
+    std::string expectedEmpty{};
+
+    SECTION("Test with null arguments list")
+    {
+        int testArgc = 0;
+        const char** testArgv = NULL;
```
Comment (ziobron) [inline bowling/tests/argumentParser.ut.cpp]:
```suggestion
        const char** testArgv = nullptr;
```
Replace it globally, everywhere

File: bowling/tests/directoryHandler.ut.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            }
+        }
+    }
+}
+
+SCENARIO("DirectoryHandler object created for directory containing one file", "FileDirectory")
+{
+    std::string testDirectoryName{"testDirectory/dirOneFile"};
+    std::vector<std::string> expectedFilesInDirectory{};
+    expectedFilesInDirectory.push_back("testDirectory/dirOneFile/file1.txt");
```
Comment (ziobron) [inline bowling/tests/directoryHandler.ut.cpp]:
```suggestion
    std::vector<std::string> expectedFilesInDirectory{"testDirectory/dirOneFile/file1.txt"};
```

File: bowling/src/fileHandler.cpp [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+{
+    if (isFileOpened()) {
+        fileStream_.close();
+    }
+}
+
+std::string FileHandler::readLine()
+{
+    if (!isFileOpened() || (fileAccess_ != FileAccess::INPUT)) {
+        return {};
```
Comment (ziobron) [inline bowling/src/fileHandler.cpp:25]: good, you can throw an exception if you wish in this case

File: bowling/tests/fileHandler.ut.cpp [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+SCENARIO("FileHandler object created for output, so lines should not be read from file", "FileHandler")
+{
+    std::string testFileName{"testDirectory/testFileWrongOutput.txt"};
+    bool expectedIsFileOpened = true;
+    std::string expectedEmpty{};
+    std::vector<std::string> expectedLinesInFile{};
+    expectedLinesInFile.push_back("line1");
+    expectedLinesInFile.push_back("line2");
+    expectedLinesInFile.push_back("line3");
+    expectedLinesInFile.push_back("");
```
Comment (ziobron) [inline bowling/tests/fileHandler.ut.cpp]: put then on the vector in constructor

File: bowling/inc/singleGame.hpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+    std::string getBowlingSigns() const { return bowlingSigns_; };
+
+    GameStatus gameStatus_{};
+
+    void setRolls(std::size_t point);
+    void setPlayerName(std::string playerName) { playerName_ = playerName; };
+    void setGameInput(std::string gameInput) { gameInput_ = gameInput; };
+    void setBowlingSigns(std::string bowlingSigns) { bowlingSigns_ = bowlingSigns; };
+    void makePointsFromSigns();
```
Comment (ziobron) [inline bowling/inc/singleGame.hpp]: I don't like the name 'signs'. 'chars' or 'characters' might be better.

File: bowling/src/singleGame.cpp [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+        return true;
+    }
+
+    if (isNotBowlingOrder()) {
+        return true;
+    }
+    return false;
+}
+
+void SingleGame::eraseSpaces()
```
Comment (ziobron) [inline bowling/src/singleGame.cpp:36]: i would merge it with setGemeInput

File: bowling/src/singleGame.cpp [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+        return true;
+    }
+
+    if (isNotAllowedChar()) {
+        return true;
+    }
+
+    if (isNotBowlingOrder()) {
+        return true;
+    }
```
Comment (ziobron) [inline bowling/src/singleGame.cpp]:
```suggestion
    return (getGameInput().empty()) ||
        isNotPlayerName() ||
        isNotBowlingSigns() ||
        isNotAllowedChar() ||
        isNotBowlingOrder());
```

File: bowling/src/singleGame.cpp [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+            }
+            else {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool SingleGame::isNotPlayerName()
```
Comment (ziobron) [inline bowling/src/singleGame.cpp:120]: I would not expect that `is*` function is not const and it changes the object state (setPlayerName). You can rethink the design of this part.
The same in `isNotBowlingSigns`. You can rename those functions.

File: bowling/src/singleGame.cpp [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+    if (foundIndexAfterName != std::string::npos) {
+        auto foundNotAllowedCharacter = gameInput.find_first_not_of(ALLOWEDCHARACTERS, ++foundIndexAfterName);
+        if (foundNotAllowedCharacter != std::string::npos) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool SingleGame::isNotBowlingOrder()
```
Comment (ziobron) [inline bowling/src/singleGame.cpp:65]: this function looks complicated. Think about refactoring it.

File: bowling/tests/singleGame.ut.cpp [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+        somePairs whosPlay{
+            {SingleGame("Romek:X|X|X|X|X|X|X|X|X|X||XX"), {10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10}},
+            {SingleGame("Zenon:9-|9-|9-|9-|9-|9-|9-|9-|9-|9-||"), {9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0}},
+            {SingleGame("Gienia:5/|5/|5/|5/|5/|5/|5/|5/|5/|5/||5"), {5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5}},
+            {SingleGame("Matylda:X|7/|9-|X|-8|8/|-6|X|X|X||81"), {10, 7, 3, 9, 0, 10, 0, 8, 8, 2, 0, 6, 10, 10, 10, 8, 1}},
+            {SingleGame("Eustachy:33|13|14|14|12|"), {3, 3, 1, 3, 1, 4, 1, 4, 1, 2}},
+            {SingleGame("Zuzka:X|13|14|14|12|"), {10, 1, 3, 1, 4, 1, 4, 1, 2}},
+            {SingleGame(":-2|13|14|14|12|"), {0, 2, 1, 3, 1, 4, 1, 4, 1, 2}}};
+
+        WHEN("Constructor indirectly call private method putScoresToVector() which analize game input, transforms input to nums and store in vector")
```
Comment (ziobron) [inline bowling/tests/singleGame.ut.cpp:50]: you don't need to mention about private methods. They can change often, so keeping it up to date may be problematic.

File: bowling/tests/singleGame.ut.cpp [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+                std::make_pair("Jan:1", "Jan:1"),
+                std::make_pair("Jan:11|", "Jan:11|"),
+                std::make_pair("Jan:1-|", "Jan:1-|"),
+                std::make_pair("Jan:12|12|12|12|12|12|12|12|12|12||", "Jan:12|12|12|12|12|12|12|12|12|12||"),
+                std::make_pair("Jan:-2|92|12|52|X|12|X|12|X|26||", "Jan:-2|92|12|52|X|12|X|12|X|26||"),
+                std::make_pair("Jan:X|7/|9-|X|-8|8/|-6|X|X|6-||8", "Jan:X|7/|9-|X|-8|8/|-6|X|X|6-||8"),
+                std::make_pair("Jan:X|7/|9-|X|-8|8/|-6|X|X|-4||X", "Jan:X|7/|9-|X|-8|8/|-6|X|X|-4||X"),
+                std::make_pair("Jan:X|X|X|X|X|X|X|X|X|X||XX", "Jan:X|X|X|X|X|X|X|X|X|X||XX"),
+                std::make_pair("Jan:--|6|--|--|--|--|--|--|--|--||", ""),
+                std::make_pair("Jan:--|--|--||--|--|--|--|--|--|--||", ""));
```
Comment (ziobron) [inline bowling/tests/singleGame.ut.cpp:234]: most of them have a correct order. They should not be under this test case

File: bowling/tests/lane.ut.cpp [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            Lane lane(dirName);
+
+            THEN("Value laneName_ is equal to filename without extension: " << expectedLaneName)
+            {
+                REQUIRE(lane.getLaneName() == expectedLaneName);
+            }
+        }
+    }
+}
```
Comment (ziobron) [inline bowling/tests/lane.ut.cpp:36]: you could also mock a FileHandler to return values not from real but provided here in the code.

File: bowling/tests/lane.ut.cpp [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+        WHEN("Lane object is creating")
+        {
+            Lane lane(dirName);
+
+            THEN("Value laneName_ is equal to filename without extension: " << expectedLaneName)
+            {
+                REQUIRE(lane.getLaneName() == expectedLaneName);
+            }
+        }
+    }
```
Comment (ziobron) [inline bowling/tests/lane.ut.cpp:35]: you should also test operator<<. You can call `.str()` on it to get its content in a string.

File: bowling/inc/bowlingGame.hpp [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+#include <string>
+#include <vector>
+
+class BowlingGame {
+public:
+    BowlingGame(int argc, const char** argv);
+    BowlingGame() = delete;
+
+    bool isHelpToBePrinted();
+    std::string& printHelp() const;
```
Comment (ziobron) [inline bowling/inc/bowlingGame.hpp]: I would rather see it outside the BowlingGame class. They may be free functions.

File: bowling/src/bowlingGame.cpp [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+void BowlingGame::outputScores()
+{
+    if (isInputDirectoryGiven()) {
+        showScores();
+
+        if (isOutputFileGiven()) {
+            saveScores();
+        }
+    }
+}
```
Comment (ziobron) [inline bowling/src/bowlingGame.cpp]: not exactly compliant with requirements. 
If output file is given we should save results into this file. If not, show them on a screen.

File: bowling/tests/bowlingGame.ut.cpp [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+                    std::stringstream ss;
+                    ss << testBowlingGame.printHelp();
+                    REQUIRE_FALSE(ss.str().empty());
+                    REQUIRE_THAT(ss.str(),
+                                 StartsWith(expectedHelpBeginning));
+                }
+            }
+        }
+    }
+}
```
Comment (ziobron) [inline bowling/tests/bowlingGame.ut.cpp:63]: what about whole application tests? You should check if proper values are printed in an output file / screen. You can refactor you showScore / saveScore functions to use some indirect function that returns an output string and check it here. 
In fact this is the only major problem in your code. The rest of my comments is really a minor thing :)
You can also use mocks to provide your own content of files.

**Review by ziobron — CHANGES_REQUESTED** [review]
I love your tests 💚
what about whole application tests? You should check if proper values are printed in an output file / screen. You can refactor you showScore / saveScore functions to use some indirect function that returns an output string and check it here. 
In fact this is the only major problem in your code. The rest of my comments is really a minor thing :)
You can also use mocks to provide your own content of files.

File: bowling/tests/bowlingGame.ut.cpp (hunk 21 above)
Comment (DamiXioR) [reply bowling/tests/bowlingGame.ut.cpp:63]: thank You for review, i tried to fix some issues, i hope it's everything allright now

File: bowling/tests/bowlingGame.ut.cpp [hunk 22]
```diff
@@ -62,3 +63,41 @@ SCENARIO("BowlingGame class can check if help is to be printed and trigger print
         }
     }
 }
+
+SCENARIO("Should save scores in file")
+{
+    GIVEN("Three arguments for start")
+    {
+        const char* userInput[] = {"./bowling", "../tests/testDirectory/goodInput/", "../tests/testDirectory/gameScores/result1.txt"};
+        BowlingGameStub bowlingGameStub(3, userInput);
```
Comment (ziobron) [inline bowling/tests/bowlingGame.ut.cpp]: why do you use stub here? what is different in stub? You want to test your original BowlingGame. Stub can be used only when you need to pass it as a dependency to check some actions. But the real thing under test should be just BowlingGame here. And if you want to replace something with stub, it should inherit from the common interface as the real object. Stub should be removed.
```suggestion
        BowlingGame bowlingGame(3, userInput);
```

File: bowling/tests/bowlingGame.ut.cpp [hunk 23]
```diff
@@ -62,3 +63,41 @@ SCENARIO("BowlingGame class can check if help is to be printed and trigger print
...
+            THEN("Should save scores in file")
+            {
+                std::vector<std::string> expectedOutput{
+                    "### lane3 : game in progress ###",
+                    "Zuzka 31",
+
+                    "### lane1 : game finished ###",
+                    "Matylda 167",
+
+                    "### lane2 : game finished ###",
```
Comment (ziobron) [inline bowling/tests/bowlingGame.ut.cpp]: why such order? why lane1 is not first?

File: bowling/tests/bowlingGame.ut.cpp [hunk 24]
```diff
@@ -62,3 +63,41 @@ SCENARIO("BowlingGame class can check if help is to be printed and trigger print
...
+                    "Pimpek 300"};
+
+                int i = 0;
+                std::string temp{};
+                std::string wholeLine{};
+
+                while (ss >> temp) {
```
Comment (ziobron) [inline bowling/tests/bowlingGame.ut.cpp]: what about ss.str()?

File: bowling/inc/bowlingGameStub.hpp [hunk 25]
```diff
@@ -0,0 +1,109 @@
+#pragma once
```
Comment (ziobron) [inline bowling/inc/bowlingGameStub.hpp]: isn't it mainly a copy of BowlingGame class? I can see a lot of code duplication.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: bowling/docs/ProjectPlan.md (hunk 2 above)
Comment (WBigaj) [reply bowling/docs/ProjectPlan.md]: Done

File: bowling/docs/ProjectPlan.md (hunk 1 above)
Comment (WBigaj) [reply bowling/docs/ProjectPlan.md:2]: Thanks :)

File: bowling/docs/plant_uml (hunk 3 above)
Comment (WBigaj) [reply bowling/docs/plant_uml:102]: :)

File: bowling/tests/bowlingGame.ut.cpp (hunk 23 above)
Comment (WBigaj) [reply bowling/tests/bowlingGame.ut.cpp]: Fixed

File: bowling/tests/bowlingGame.ut.cpp (hunk 22 above)
Comment (WBigaj) [reply bowling/tests/bowlingGame.ut.cpp]: Done :+1:

File: bowling/inc/bowlingGameStub.hpp (hunk 25 above)
Comment (WBigaj) [reply bowling/inc/bowlingGameStub.hpp]: Removed redundant Stub.

File: bowling/tests/bowlingGame.ut.cpp (hunk 24 above)
Comment (WBigaj) [reply bowling/tests/bowlingGame.ut.cpp]: Changed to work with BowlingClass and used str().

**Review by ziobron — APPROVED** [review]
Well done 🥇

**Comment by ziobron** [comment]
{
"requirement_ids": [4595, 4595, 4595, 4595, 4595, 4595, 4595, 4596, 1120],
"points": [10, 10, 10, 10, 10, 10, 10, 20, 5],
"users": ["NG90", "DamiXioR", "WBigaj", "Marimeiko"]
}

**Pull request closed without merging**